### PR TITLE
OSAC-3082: implement reference validation interceptor

### PR DESCRIPTION
**Pull request by htayrie-rh**
## [OSAC-3082](https://redhat.atlassian.net/browse/OSAC-3082): Implement reference validation interceptor

**Jira:** [OSAC-3082](https://issues.redhat.com/browse/OSAC-3082)
**Story type:** [DEV]

### Summary

Implements a centralized gRPC unary server interceptor that validates resource references in Create and Update requests using protoreflect. The interceptor discovers reference-typed fields by naming convention (`*Reference`, `*LocalReference`), validates referenced resources exist via registered lookup functions, auto-populates missing id/name fields, and returns aggregated `InvalidArgument` errors with structured `BadRequest` `FieldViolation` details.

This is the foundation for [OSAC-1330](https://redhat.atlassian.net/browse/OSAC-1330) (type-safe resource references). The interceptor is registered in the gRPC chain but has no lookup functions yet — it becomes active when subsequent epics (Networking, Compute, IP Management) register their reference types.

### Changes

**New package: `internal/references/`**
- `ReferenceValidator` interceptor with builder pattern (`NewReferenceValidator().SetLogger().SetMetricsRegisterer().Build()`)
- Protoreflect-based message walker handling singular, repeated, nested, and oneof fields
- Three resolution modes: name-only (auto-populate id), id-only (auto-populate name), both-provided (verify agreement)
- Tenant context extraction: local refs use owning resource's tenant/project; full refs support explicit project and shared tenant
- Fail-closed design: unregistered reference types return `Internal` error
- Prometheus metrics: `osac_reference_validation_total` (CounterVec) and `osac_reference_validation_duration_seconds` (HistogramVec)

**Interceptor chain registration**
- Added after JIT provisioning interceptor, before server handlers
- Both unary (validates) and stream (pass-through) chains

**Test proto definitions**
- `proto/tests/osac/tests/v1/reference_test_types.proto` with 7 message types and 3 request types covering all reference field patterns

### Testing
- **Unit tests:** 45 Ginkgo tests covering builder, method filtering, field discovery (simple/nested/repeated/oneof), stream pass-through, reference type detection, lookup registration, tenant context extraction, error reporting/aggregation, non-proto pass-through, and Prometheus metrics
- **Integration tests:** N/A (interceptor is inert until lookup functions are registered in subsequent epics)
- **Coverage:** 84.6% — all public API behavioral contracts tested

### Acceptance Criteria

- [x] AC-1: gRPC unary server interceptor validates reference-typed fields in Create and Update requests
- [x] AC-2: Detects reference fields by proto message naming convention via protoreflect
- [x] AC-3: Invalid references return `InvalidArgument` with `BadRequest` containing `FieldViolation` entries with field paths
- [x] AC-4: Aggregates all reference errors in a single response (no short-circuit)
- [x] AC-5: Local references use owning resource's tenant/project; full references use explicit tenant/project with fallback
- [x] AC-6: Prometheus metrics emitted (`osac_reference_validation_total`, `osac_reference_validation_duration_seconds`)

## Summary by CodeRabbit

* **New Features**
  * Added automatic gRPC validation for protobuf “reference” fields on unary Create/Update requests, including nested, repeated, one-of, and local/shared variants.
  * When possible, missing reference identifiers (ID/name) are auto-populated after successful lookup; invalid or conflicting references return field-level `InvalidArgument` errors.
  * Validation emits Prometheus metrics for valid/invalid/error outcomes and lookup timing.
* **Tests**
  * Added an extensive Ginkgo/Gomega suite plus new protobuf test types to cover success, failure, aggregation, and metric behavior.
* **Chores**
  * Promoted select Go dependencies from indirect to direct requirements.

### CODE REVIEW

File: internal/cmd/service/start/grpcserver/start_grpc_server_cmd.go [hunk 1]
```diff
@@ -430,6 +431,16 @@ func (c *runnerContext) run(cmd *cobra.Command, argv []string) error { //nolint:
...
+	// Prepare the reference validation interceptor:
+	c.logger.InfoContext(ctx, "Creating reference validation interceptor")
+	referenceValidator, err := references.NewReferenceValidator().
+		SetLogger(c.logger).
+		SetMetricsRegisterer(metricsRegisterer).
+		Build()
+	if err != nil {
+		return fmt.Errorf("failed to create reference validation interceptor: %w", err)
+	}
+
```
[the remark this answers is not in the document]
Comment (htayrie-rh) [reply internal/cmd/service/start/grpcserver/start_grpc_server_cmd.go:443]: No production proto messages use `*Reference` type names — only test protos in `proto/tests/`. The interceptor is intentionally inert until subsequent epics (OSAC-3093 Networking, etc.) define reference proto messages and register their lookup functions. The fail-closed behavior is a safety net for when those registrations are added, not a current risk.

File: internal/cmd/service/start/grpcserver/start_grpc_server_cmd.go [hunk 2]
```diff
@@ -504,6 +516,7 @@ func (c *runnerContext) run(cmd *cobra.Command, argv []string) error { //nolint:
 			authnInterceptor.StreamServer,
 			authzInterceptor.StreamServer,
 			jitProvisioningInterceptor.StreamServer,
+			referenceValidator.StreamServer,
```
[the remark this answers is not in the document]
Comment (htayrie-rh) [reply internal/cmd/service/start/grpcserver/start_grpc_server_cmd.go:519]: Every interceptor in the codebase registers both `UnaryServer` and `StreamServer` — removing it would be the only exception. The overhead of one pass-through closure is negligible, and having the hook in the chain avoids a follow-up PR when stream validation is needed.

File: internal/references/reference_validator_test.go [hunk 3]
```diff
@@ -0,0 +1,1377 @@
...
+					}.Build(),
+					Spec: testsv1.TestRefSpec_builder{}.Build(),
+				}.Build(),
+			}.Build()
+
+			refs := discoverReferenceFields(request)
+
+			Expect(refs).To(BeEmpty())
+		})
+	})
```
[the remark this answers is not in the document]
Comment (htayrie-rh) [reply internal/references/reference_validator_test.go:457]: Valid concern about testing through a mirror walker. The discovery tests verify shape coverage (singular, nested, repeated, oneof) while the production walker is tested end-to-end through the resolution tests. The `TextName()` vs `Name()` divergence is cosmetic — both return identical values for proto3.

That said, if `walkMessage` broke specifically for nested-repeated fields, neither test path would catch it independently. Acknowledged as a real gap but low-probability — deferring the rewrite for now.

File: internal/references/reference_validator_test.go [hunk 4]
```diff
@@ -0,0 +1,1377 @@
...
+				&grpc.UnaryServerInfo{FullMethod: "/osac.tests.v1.TestService/Create"},
+				mockHandler,
+			)
+
+			Expect(err).To(HaveOccurred())
+			status, ok := grpcstatus.FromError(err)
+			Expect(ok).To(BeTrue())
+			Expect(status.Code()).To(Equal(grpccodes.InvalidArgument))
+			Expect(status.Message()).To(ContainSubstring("do not refer to the same resource"))
+		})
```
[the remark this answers is not in the document]
Comment (htayrie-rh) [reply internal/references/reference_validator_test.go:1151]: Fixed in 40acd5f4 — added a test where both id and name are provided and match the resolved values, asserting the handler executes and no error is returned.

File: internal/references/reference_validator_test.go [hunk 5]
```diff
@@ -0,0 +1,1377 @@
...
+			mockHandler := func(ctx context.Context, req any) (any, error) {
+				return nil, nil
+			}
+
+			_, _ = validator.UnaryServer(
+				context.Background(),
+				request,
+				&grpc.UnaryServerInfo{FullMethod: "/osac.tests.v1.TestService/Create"},
+				mockHandler,
+			)
```
[the remark this answers is not in the document]
Comment (htayrie-rh) [reply internal/references/reference_validator_test.go]: Fixed in 40acd5f4 — metric tests now assert the gRPC status code alongside the counter value.

File: internal/references/reference_validator.go [hunk 6]
```diff
@@ -0,0 +1,488 @@
...
+func (v *ReferenceValidator) walkMessage(ctx context.Context, msg protoreflect.Message, path []string,
+	violations *[]*errdetails.BadRequest_FieldViolation, tenant, project string) error {
+	var internalErr error
+
+	msg.Range(func(fd protoreflect.FieldDescriptor, val protoreflect.Value) bool {
+		if fd.Kind() != protoreflect.MessageKind {
+			return true
+		}
+
+		fieldPath := append(append([]string{}, path...), string(fd.Name()))
```
[the remark this answers is not in the document]
Comment (htayrie-rh) [reply internal/references/reference_validator.go:214]: Fixed in 40acd5f4 — violations are now sorted by `Field` before serializing the `BadRequest` detail.

File: internal/references/reference_validator.go [hunk 7]
```diff
@@ -0,0 +1,488 @@
...
+
+		if fd.IsMap() {
+			return true
+		}
```
[the remark this answers is not in the document]
Comment (htayrie-rh) [reply internal/references/reference_validator.go:223]: Fixed in 40acd5f4 — map fields with message values now log a warning instead of being silently skipped. Full map walking deferred until a production use case exists (no current protos use map fields with reference values).

File: internal/references/reference_validator.go [hunk 8]
```diff
@@ -0,0 +1,488 @@
...
+			"reference type %q does not have required id/name fields", fullName)
+	}
+	id := refMsg.Get(idField).String()
+	name := refMsg.Get(nameField).String()
+
+	resourceType := string(fullName.Name())
+	fieldPath := strings.Join(path, ".")
+
+	start := time.Now()
+	resolved, err := lookupFunc(ctx, refTenant, refProject, id, name)
```
[the remark this answers is not in the document]
Comment (htayrie-rh) [reply internal/references/reference_validator.go:316]: Fixed in 40acd5f4 — empty references (both id and name blank) now short-circuit into a violation without calling the lookup. In practice, `buf.validate` CEL requires `id_or_name_required` on all reference types and protovalidate runs earlier in the chain, but defense in depth is cheap here.

File: internal/references/reference_validator.go [hunk 9]
```diff
@@ -0,0 +1,488 @@
...
+	elapsed := time.Since(start)
+
+	v.recordDuration(resourceType, elapsed)
```
[the remark this answers is not in the document]
Comment (htayrie-rh) [reply internal/references/reference_validator.go:319]: Serial lookups are by design — they share the DB transaction established by the transaction interceptor earlier in the chain. Concurrent lookups would require separate connections and complicate transaction isolation. Typical requests have 1-3 references. If performance becomes a concern with high-cardinality repeated reference fields, we can add concurrency and bounds then — premature optimization here would change the concurrency model.

File: internal/references/reference_validator.go [hunk 10]
```diff
@@ -0,0 +1,488 @@
...
+		return grpcstatus.Errorf(grpccodes.Internal,
+			"internal error resolving reference at %s", fieldPath)
+	}
+
+	if id == "" && resolved.ID != "" {
+		refMsg.Set(idField, protoreflect.ValueOfString(resolved.ID))
+	}
+	if name == "" && resolved.Name != "" {
+		refMsg.Set(nameField, protoreflect.ValueOfString(resolved.Name))
+	}
```
[the remark this answers is not in the document]
Comment (htayrie-rh) [reply internal/references/reference_validator.go:363]: Fixed in 40acd5f4 — nil `resolved` with nil error is now caught and returns an `Internal` error with a diagnostic log.

File: internal/references/reference_validator.go [hunk 11]
```diff
@@ -0,0 +1,488 @@
...
+				Field:       fieldPath,
+				Description: desc,
+			})
+			v.logger.WarnContext(ctx, "Reference not found",
+				"field_path", fieldPath,
+				"reference_type", resourceType,
+				"identifier", identifier(id, name),
+				"tenant", refTenant,
+				"project", refProject,
+			)
```
[the remark this answers is not in the document]
Comment (htayrie-rh) [reply internal/references/reference_validator.go:334]: Fixed in 40acd5f4 — customer-scoped slog keys (`identifier`, `tenant`, `project`, `id`, `name`) now use the `!` prefix for redaction, consistent with `internal/auth/`.

File: internal/references/reference_validator.go [hunk 12]
```diff
@@ -0,0 +1,488 @@
...
+		tenant = metadataMsg.Get(tenantField).String()
+	}
+
+	projectField := metadataMsg.Descriptor().Fields().ByName("project")
+	if projectField != nil {
+		project = metadataMsg.Get(projectField).String()
+	}
+
+	return
+}
```
[the remark this answers is not in the document]
Comment (htayrie-rh) [reply internal/references/reference_validator.go:463]: The interceptor runs at position 9 in the chain — after Auth (position 6) and Authz/OPA (position 7). By this point, OPA has already authorized the caller's access to the tenant in the request metadata. Deriving tenant from request metadata is the intended design per the enhancement proposal.

The only "tenant override" is `shared=true`, which maps to the platform-level shared tenant — not an arbitrary tenant. This is by design: shared resources (NetworkClass, PublicIPPool) are meant to be visible to all tenants. The explicit `project` field allows cross-project references within the same tenant, also by design.

There is no mechanism for a caller to point resolution at an arbitrary tenant.

File: internal/references/reference_validator.go [hunk 13]
```diff
@@ -0,0 +1,488 @@
...
+	if err != nil {
+		var registered prometheus.AlreadyRegisteredError
+		if errors.As(err, &registered) {
+			return registered.ExistingCollector.(C), nil
+		}
+		var zero C
+		return zero, err
+	}
+	return c, nil
+}
```
[the remark this answers is not in the document]
Comment (htayrie-rh) [reply internal/references/reference_validator.go:531]: Fixed in 40acd5f4 — `registerOrReuse` now uses the comma-ok type assertion form.

File: internal/references/reference_validator.go [hunk 14]
```diff
@@ -0,0 +1,521 @@
...
+		return grpcstatus.Errorf(grpccodes.Internal,
+			"reference type %q does not have required id/name fields", fullName)
+	}
+	id := refMsg.Get(idField).String()
+	name := refMsg.Get(nameField).String()
+
+	resourceType := string(fullName.Name())
+	fieldPath := strings.Join(path, ".")
+
+	if id == "" && name == "" {
```
Comment (danielerez) [inline internal/references/reference_validator.go:306]: Maybe worth adding a test also for this scenario (empty references)

File: internal/references/reference_validator.go [hunk 15]
```diff
@@ -0,0 +1,521 @@
...
+		validationTotal:    validationTotal,
+		validationDuration: validationDuration,
+	}
+	return
+}
+
+// Register associates a reference message type with a lookup function. Must be called before the
+// gRPC server starts accepting requests.
+func (v *ReferenceValidator) Register(fullName protoreflect.FullName, lookupFunc ReferenceLookupFunc) {
+	v.registry[fullName] = lookupFunc
```
Comment (danielerez) [inline internal/references/reference_validator.go:138]: Do we need some mechanism to ensure that Register indeed "called before the gRPC server starts accepting requests"?
E.g. we can use some latch as atomic.Bool to avoid register more than once.

**Review by danielerez — COMMENTED** [review]
Looks good, just a couple of comments.

File: internal/references/reference_validator.go [hunk 16]
```diff
@@ -0,0 +1,521 @@
...
+						internalErr = err
+						return false
+					}
+				} else {
+					err := v.walkMessage(ctx, elemMsg, indexedPath, violations, tenant, project)
+					if err != nil {
+						internalErr = err
+						return false
+					}
+				}
```
Comment (rccrdpccl) [inline internal/references/reference_validator.go:246]: nit: else condition is never necessary. rewriting it in a way to eliminate it will simplify the nesting, make it easier to understand by humans and give good examples patterns for the AI to replicate. Commenting only on this one but there are other instances as well

File: internal/references/reference_validator.go (hunk 15 above)
Comment (htayrie-rh) [reply internal/references/reference_validator.go:138]: Good point. The current architecture guarantees ordering — `Register` is called during server construction in `start_grpc_server_cmd.go`, sequentially, before `grpc.Serve()` starts accepting connections. So there's no concurrent access risk today.

That said, adding a freeze mechanism (e.g. a `frozen bool` that `Build()` or a `Freeze()` method sets, with `Register` panicking after) would make the contract explicit and guard against future misuse. Do you think it's worth adding now, or is the startup ordering sufficient for the current codebase?

Comment (rccrdpccl) [reply internal/references/reference_validator.go:138]: I agree, even if the actual risk is minimal we should do this to establish a good pattern

File: internal/references/reference_validator.go [hunk 17]
```diff
@@ -0,0 +1,521 @@
...
+		)
+		v.recordResult(resourceType, "invalid")
+		return nil
+	}
+
+	v.logger.DebugContext(ctx, "Reference validated",
+		"field_path", fieldPath,
+		"reference_type", resourceType,
+		"resolved_id", resolved.ID,
+		"resolved_name", resolved.Name,
```
Comment (rccrdpccl) [inline internal/references/reference_validator.go]: shouldn't we use ! prefix here too?

**Review by rccrdpccl — CHANGES_REQUESTED** [review]

File: internal/references/reference_validator_test.go [hunk 18]
```diff
@@ -0,0 +1,1430 @@
...
+						Attachment: testsv1.TestRefAttachment_builder{
+							Subnet: testsv1.TestTargetLocalReference_builder{
+								Name: "my-subnet",
+							}.Build(),
+						}.Build(),
+					}.Build(),
+				}.Build(),
+			}.Build()
+
+			refs := discoverReferenceFields(request)
```
Comment (rccrdpccl) [inline internal/references/reference_validator_test.go:301]: Why aren't we testing nested paths through UnaryServer? Wouldn't that be a more complete test?

Comment (rccrdpccl) [reply internal/references/reference_validator_test.go:301]: That' be for nested, nested repeated, OneOf

File: internal/references/reference_validator.go (hunk 14 above)
Comment (htayrie-rh) [reply internal/references/reference_validator.go:306]: Good catch — added a test for the empty reference case (both id and name empty). It verifies the lookup function is never called and the error contains "must specify id or name".

File: internal/references/reference_validator.go (hunk 16 above)
Comment (htayrie-rh) [reply internal/references/reference_validator.go:246]: Good call — fixed both instances in `walkMessage`. Used `continue` in the list loop and `return true` in the Range callback to eliminate the else nesting.

File: internal/references/reference_validator.go (hunk 15 above)
Comment (danielerez) [reply internal/references/reference_validator.go:138]: > Good point. The current architecture guarantees ordering — `Register` is called during server construction in `start_grpc_server_cmd.go`, sequentially, before `grpc.Serve()` starts accepting connections. So there's no concurrent access risk today.
> 
> That said, adding a freeze mechanism (e.g. a `frozen bool` that `Build()` or a `Freeze()` method sets, with `Register` panicking after) would make the contract explicit and guard against future misuse. Do you think it's worth adding now, or is the startup ordering sufficient for the current codebase?

maybe something simple as:
```
func (v *ReferenceValidator) Register(...
    if v.sealed.Load() {
        panic(fmt.Sprintf("Register called after interceptor started serving: %s", fullName))
    }
    v.registry[fullName] = lookupFunc
```

and seal when the server is active. e.g. in UnaryServer func:
```
func (v *ReferenceValidator) UnaryServer(...
    v.sealed.Store(true)
```

File: internal/references/reference_validator.go (hunk 17 above)
Comment (htayrie-rh) [reply internal/references/reference_validator.go]: Fixed — added `!` prefix to both `resolved_id` and `resolved_name`.

File: internal/references/reference_validator.go (hunk 15 above)
Comment (htayrie-rh) [reply internal/references/reference_validator.go:138]: Good suggestion — implemented it. Added `sealed atomic.Bool` to the struct, `v.sealed.Store(true)` at the top of `UnaryServer`, and a panic guard in `Register`. Also added a test that verifies the panic fires after the first request.

File: internal/references/reference_validator_test.go (hunk 18 above)
Comment (htayrie-rh) [reply internal/references/reference_validator_test.go:301]: Fair point — added an end-to-end `UnaryServer` test that builds the complex message with all reference patterns (singular full ref, singular local ref, nested ref in attachment, repeated refs in both attachment and top-level), registers lookups for all 3 types, and asserts all 5 references get their `id` auto-populated through the real interceptor.

File: internal/references/reference_validator.go [hunk 19]
```diff
@@ -0,0 +1,527 @@
...
+	refTenant, refProject := resolveTenantProject(refMsg, fullName, callerTenant, callerProject)
+
+	idField := refMsg.Descriptor().Fields().ByName("id")
+	nameField := refMsg.Descriptor().Fields().ByName("name")
+	if idField == nil || nameField == nil {
+		return grpcstatus.Errorf(grpccodes.Internal,
+			"reference type %q does not have required id/name fields", fullName)
+	}
+	id := refMsg.Get(idField).String()
+	name := refMsg.Get(nameField).String()
```
[the remark this answers is not in the document]
Comment (htayrie-rh) [reply internal/references/reference_validator.go:301]: fixed
